websocket.go

```diff
@@ -193,25 +193,45 @@ func (t wsHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
 		s := t.service
 		var reply []byte
+		var tun *StreamingTunnel
 		path := strings.TrimPrefix(r.URL.Path, "/"+t.serviceName+"/")
 		log.Lvlf2("ws request from %s: %s/%s", r.RemoteAddr, t.serviceName, path)
-		reply, err = s.ProcessClientRequest(r, path, buf)
+		reply, tun, err = s.ProcessClientRequest(r, path, buf)
 		if err == nil {
-			tx += len(reply)
-			err := ws.WriteMessage(mt, reply)
-			if err != nil {
-				log.Error(err)
-				return
+			if tun == nil {
+				tx += len(reply)
+				err := ws.WriteMessage(mt, reply)
+				if err != nil {
+					log.Error(err)
+					break
+				}
+			} else {
+				for {
+					select {
+					case reply, ok := <-tun.out:
+						if !ok {
+							err = errors.New("service finished streaming")
+							close(tun.close)
+							break outerReadLoop
+						}
+						tx += len(reply)
+						err = ws.WriteMessage(mt, reply)
```

Comment:
I was expecting to see something about Write deadlines here.

Reply:
I see that gorilla websockets has `Set{Read,Write}Deadline`, I could use that. But what would be the appropriate deadline, a minute? Assuming it's to stop client/server from sending unrealistically large messages.

Reply:
It is to stop a badly behaved client from trickling a message in a byte at a time, holding open server resources. Same for reading... for example when a server has a torrent of messages and the client just stops talking TCP (phone out of GSM range, out of battery, etc) then all those messages sit in kernel buffers until TCP times out.